src/qutip_tensornetwork/core/data/network.py

```diff
@@ -672,3 +754,108 @@ def _match_edges_by_split(out_edges, in_edges):
     new_out_edges.reverse()
     new_in_edges.reverse()
     return new_out_edges, new_in_edges
+
+
+def _match_dimensions(edges, target_dims):
+    """This function splits the edges such that the dimensions of edges
+    matches target_dims. Only split of edges are allowed in order to match
+    target_dims.
+
+    Note
+    ----
+    The modifications to edges are performed in-place, which is why the
+    function is private as it is meant to be used only within network.
+
+    Examples
+    --------
+    >>>network_dim = [4]
+    >>>target_dims = [2, 2]
+    >>>array = np.random.random(network_dim)
+    >>>node = tn.Node(array)
+    >>>network = Network(node[:], [])
+    >>>_match_dimensions(network.in_edges, target_dims)
+    ValueError
+
+    Since only splitting of edges are allowed, the next example will raise a
+    ValueError.
+    >>>network_dim = [2, 2]
+    >>>target_dims = [4]
+    >>>array = np.random.random(network_dim)
+    >>>node = tn.Node(array)
+    >>>network = Network(node[:], [])
+    >>>_match_dimensions(network.in_edges, target_dims)
+    ValueError
+    """
+
+    edges = edges[:]
+    target_dims = target_dims[:]
+    _target_dims = target_dims[:]  # This is kept for the error messages.
+
+    e_dims = [e.dimension for e in edges]
+
+    new_edges = []
+
+    if len(edges) == 0 and len(target_dims) == 0:
+        return _edges
+
+    if len(edges) == 0 or len(target_dims) == 0:
+        raise ValueError(
+            "edges are not compatible. The dimensions for edges is "
+            + str(e_dims)
+            + ", whereas the target dimension is"
+            + str(_target_dims)
+        )
+
+    if np.prod(e_dims) != np.prod(target_dims):
+        raise ValueError(
+            "edges are not compatible. The dimensions of edges is "
+            + str(e_dims)
+            + ", whereas the target dimension is "
+            + str(_target_dims)
+        )
+
+    edge = edges.pop()
+    target_dim = target_dims.pop()
+
+    while True:
+        if edge.dimension == target_dim:
+            new_edges.append(edge)
+
+            if len(edges) == 0 and len(target_dims) == 0:
+                break
+
+            edge = edges.pop()
+            target_dim = target_dims.pop()
+
+        elif edge.dimension > target_dim:  # Split edge
+            if edge.dimension % target_dim != 0:
```

Comment:
When `ValueError` is raised here the edge list is a mess. This means someone could call `qobj.ptrace()` and then when an exception is raised the Network referenced by `qobj._data` is in an invalid state. There are two good options -- copy all the edges at the start (which has some downsides) or check that everything will succeed at the start (probably with a similar loop that doesn't call `tn.split_edge`).

With the second approach, one can probably keep a record of the needed edges and splits and just apply those in a simpler loop afterwards.

Reply:
I am going to go with the copy approach just because it is simpler. We may want to revisit this in the future though.